File: zenscript-code-model/src/main/java/raylras/zen/model/type/IntersectionType.java
package raylras.zen.model.type;

import raylras.zen.model.symbol.Symbol;
import raylras.zen.model.symbol.SymbolProvider;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public record IntersectionType(List<Type> typeList) implements Type, SymbolProvider {

    @Override
    public String getTypeName() {
        return typeList.stream().map(Type::getTypeName).collect(Collectors.joining(" & "));
    }

    @Override
    public boolean isSuperclassTo(Type type) {
        return typeList.stream().anyMatch(it -> it.isSuperclassTo(type));
    }

    @Override
    public Collection<Symbol> getSymbols() {
        MemberValidator validator = new MemberValidator();
        for (Type type : typeList) {
            if (type instanceof SymbolProvider provider) {
                validator.addAll(provider.getSymbols());
            }
        }
        return validator.getMembers();
    }

    @Override
    public String toString() {
        return getTypeName();
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/type/IntersectionType.kt
package raylras.zen.model.type

import raylras.zen.model.CompilationEnvironment
import raylras.zen.model.symbol.Symbol
import raylras.zen.model.symbol.SymbolProvider

data class IntersectionType(val typeList: List<Type>) : Type, SymbolProvider {
    override val typeName: String by lazy {
        typeList.joinToString(" & ") { it.typeName }
    }

    override val simpleTypeName: String by lazy {
        typeList.joinToString(" & ") { it.simpleTypeName }
    }

    override fun isSupertypeTo(type: Type): Boolean {
        return typeList.any { it.isSupertypeTo(type) }
    }

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        val validator = MemberValidator()
        for (type in typeList) {
            if (type is SymbolProvider) {
                type.getSymbols().forEach { validator.add(it) }
            }
        }
        return validator.getMembers().asSequence()
    }

    override fun toString() = typeName
}
